book/html.md

```diff
@@ -142,66 +224,130 @@ sibling to [troff][troff], now used for Linux man pages. The
 [troff]: https://troff.org
 [jtc1-sc34]: https://www.iso.org/committee/45374.html
 
+Debugging a parser
+==================
 
-Self-closing tags
-=================
+How do we know our parser does the right thing---that it builds the
+right tree? Well the place to start is *seeing* the tree it produces.
+We can do that with a quick, recursive pretty-printer:
+
+``` {.python}
+def print_tree(node, indent=0):
+    print(" " * indent, node)
+    for child in node.children:
+        print_tree(child, indent + 2)
+```
 
-Try running this parser on this page, and you'll find that `parse`
-doesn't return anything; let's find out why:
+Here we're printing each node in the tree, and using indentation to
+show the tree structure. Since we need to print each node, it's worth
+taking the time to give them a nice printed form, which in Python
+means defining the `__repr__` function:
 
 ``` {.python expected=False}
-def parse(tokens):
-    # ...
-    print(currently_open)
-    raise Exception("Reached last token before end of document")
+class Text:
+    def __repr__(self):
+        return repr(self.text)
+
+class Element:
+    def __repr__(self):
+        return "<" + self.tag + ">"
 ```
 
-Python prints a list of `ElementNode` objects, meaning that there were
-open HTML elements still around when it reached the last token:
+Try this out on this web page, parsing the HTML source code and then
+calling `print_tree` to visualize it:
 
+``` {.python expected=False}
+headers, body = request(sys.argv[1])
+nodes = HTMLParser(body).parse()
+print_tree(nodes)
 ```
-[<__main__.ElementNode object at 0x101399c70>,  ...]
+
+Run it on this web page, and you'll see something like this:
+
+``` {.example}
+ <!DOCTYPE html >
```

Comment:
Change code to get rid of these extra spaces before the >; left another comment in the code file for this as well.